Use ActiveRecord-esque Models for RMAPI Resources

sadatay opened the pull request:
The intent of this work is to standardize all our models according to the conventions of ActiveRecord/ActiveModel.  To do this I created a base class for models that pull ActiveModel behavior in, as well as request brokerage via [flexirest](https://github.com/flexirest/flexirest) which buys us a clean DSL and ETag caching.

As much as possible has been pushed to the serialization layers.  Models now behave like AR models, i.e. `Package.find(params[:id])`, `Package.all`, etc.  All inclusion relationships have been wired up and tested.

**Included Resources / Relationships**
- Vendor Search / Detail `/eholdings/jsonapi/vendors/(:vendorId)`
  - `has_many` Packages `?include=packages`
- Package Search / Detail `/eholdings/jsonapi/packages/(:vendorId-:packageId)`
  - `has_one` Vendor `?include=vendor`
  - `has_many` Customer Resources `?include=customerResources`
- Title Search / Detail `/eholdings/jsonapi/titles/(:titleId)`
  - `has_many` Customer Resources `?include=customerResources`
- Customer Resource Detail `/eholdings/jsonapi/packages/(:vendorId-:packageId-:titleId)`
  - `has_one` Vendor `?include=vendor`
  - `has_one` Title `?include=title`
  - `has_one` Package `?include=package`

**Added 11/22**
- Given the issues @cherewaty and myself ran into yesterday with RMAPI sandbox, we converted the whole test suite to run against production Okapi and production RMAPI by default.  
  - We pulled variables that are common to all request tests (`TEST_OKAPI_TOKEN`, `headers`, etc.) out into a shared context helper and exposed the url and tenant as `TEST_OKAPI_{TENANT, URL}`, which will come in handy when this is passed off to Folio org.  
  - VCR cassettes were re-recorded on production data and specs were tweaked to account for new data.
- With a passing test suite in hand we were able to implement the `around_action` thing with confidence and clean up all the controllers.

**Added 11/27**
- New work from @cowboyd moves RMAPI configuration to the model's instance level, and sidesteps some limitations of Felxirest.  

Closes #26

Commit:
Use ActiveRecord-esque models for RMAPI Resources

This is a refactor of our previous JSONAPI work, as well as some new functionality.  The goal is to push all of the data loading from RMAPI into the model layer, thus substantially slimming down the controllers.  With this work, all RMAPI resources are now accessible via JSONAPI endpoints, and all relationships between the resources have been wired up.  Writes are not yet supported, but this is intended to be a feature-complete implementation of all the read operations needed by the frontend.

sadatay commented on ModuleDescriptor.json:
```diff
@@ -26,6 +26,10 @@
         },
+        {
+          "methods": [ "GET" ],
+          "pathPattern": "/eholdings/jsonapi/customer-resources*"
```
if the resource type is `customerResources` maybe we should use camelCase here too, as much as it pains me.

cowboyd replied:
Nooooooooooooooooo.

sadatay commented on app/controllers/customer_resources_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      )
+      render jsonapi: customer_resource,
+             include: params[:include]
+    rescue Flexirest::HTTPClientException, Flexirest::HTTPServerException, Flexirest::HTTPNotFoundClientException => e
```
this begin/rescue is cumbersome and pops up all over the place.  best thing i could think of is to implement methods in the superclass that wrap a `yield`, but that's a little heavy so I didn't want to do it without fielding other opinions.

cowboyd replied:
You can use an `around_action` for just this purpose.

Also every method gets it's own implicit `begin`/`end` block, so you can just have the `rescue` alone.

cowboyd commented on Gemfile.lock:
```diff
@@ -63,7 +63,14 @@ GEM
     ffi (1.9.18)
+    flexirest (1.5.3)
+      activesupport
+      crack
```
I'm amazed that this gem is still around.

cowboyd reviewed (commented):
Flexirest is pretty amazing.... However, rather than explicitly configuring the global flexirest resources with each request, let's try subclassing them like so:

```ruby
def customer_resources
  Class.new(CustomerResource).tap do |cls|
    cls.send(:define_method, :config) { config }
  end
end
```

cowboyd committed:
Remove superflous begin/end expression

In Ruby, every method has its own implicit exception handling block,
and so `rescue` expressions can be directly integrated into a method.

This removes the surrounding, explicity `begin/end` and just uses the
implicit method `rescue`.

cowboyd committed:
Scope configuration per-request.

There can be multiple requests happening simultaneously for different
tenants, and so we don't want to mutate the global class constants for
our flexirest resources. E.g. if `Title.configure` has side-effects,
then it could mess with another request in the same process.

This makes it so that each request gets its own subclass of a resource
configured just for that request, so no conflicts between tenants.
